apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SimpleSubscriber.cs

```diff
@@ -748,49 +780,39 @@ private class SingleChannel
-                        while (_ackQueue.Count > 0 && acks.Count < _maxAckExtendCount)
+                        if (_softStopCts.IsCancellationRequested && _qLock.Locked(() => _unAckedMsgCount == 0))
                         {
-                            acks.Add(_ackQueue.Dequeue());
+                            // All acks sent and a stop has been requested, so cancel local _hardStopCts
+                            // to signal this SingleChannel has completed; then exit Pusher
+                            _hardStopCts.Cancel();
+                            return;
                         }
-                        // Send extends that are waiting, but no more acks+extends than _maxAckExtendCount
-                        while (_extendQueue.Count > 0 && (acks.Count + extends.Count) < _maxAckExtendCount)
+                        var isCancelled = await _taskHelper.ConfigureAwaitHideCancellation(() => _qEvent.WaitAsync(streamingPullToken));
+                        if (isCancelled)
                         {
-                            extends.Add(_extendQueue.Dequeue());
+                            return;
                         }
                     }
-                    // If anything to be sent, then send it now
-                    if (acks.Count > 0 || extends.Count > 0)
+                    // Get local copies of acks/extensions to send.
+                    var acks = _qLock.Locked(() => _ackQueue.Take(_maxAckExtendCount).ToList());
+                    var extends = _qLock.Locked(() => _extendQueue.Take(_maxAckExtendCount - acks.Count).ToList());
+                    // Send acks/extensions
+                    await _taskHelper.ConfigureAwait(pull.WriteAsync(new StreamingPullRequest
                     {
-                        await _taskHelper.ConfigureAwait(pull.WriteAsync(new StreamingPullRequest
-                        {
-                            AckIds = { acks },
-                            ModifyDeadlineAckIds = { extends },
-                            ModifyDeadlineSeconds = { extends.Select(_ => _modifyDeadlineSeconds) },
-                        }));
-                        // Decrement _unAckedMsgCount by the number of acks just successfully sent.
-                        _qLock.Locked(() => _unAckedMsgCount -= acks.Count);
-                    }
-                    if (_softStopCts.IsCancellationRequested && _qLock.Locked(() => _unAckedMsgCount) == 0)
+                        AckIds = { acks },
+                        ModifyDeadlineAckIds = { extends },
+                        ModifyDeadlineSeconds = { Enumerable.Repeat(_modifyDeadlineSeconds, extends.Count) },
+                    }));
+                    // If the send didn't have errors, then dequeue the acks/extends that were just sent.
+                    // Any error would have thrown an exception in the previous WriteAsync() call.
+                    lock (_qLock)
```

Comment:
Why does this use a lock instead of _qLock.Locked?

Reply:
Functionally they're identical. _qLock.Locked() is generally used when the lock is returning something, so it saves quite a bit of code.